### PR TITLE
feat: support multiple origins in webauthn config

### PR DESCRIPTION
**Pull request by hperl**
## Related issue(s)

https://github.com/ory/docs/pull/1474

## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [ ] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [ ] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [ ] I have read the [security policy](../security/policy).
- [ ] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

File: selfservice/flow/recovery/handler.go [hunk 1]
```diff
@@ -430,6 +431,7 @@ func (h *Handler) updateRecoveryFlow(w http.ResponseWriter, r *http.Request, ps
 		} else if errors.Is(err, flow.ErrCompletedByStrategy) {
 			return
 		} else if err != nil {
+			fmt.Println(err)
```
Comment (Benehiko) [inline selfservice/flow/recovery/handler.go:434]: remove?

**Review by Benehiko — CHANGES_REQUESTED** [review]
Looks good, how will this affect current Webauthn credentials?

File: driver/config/config.go [hunk 2]
```diff
@@ -1364,11 +1365,13 @@ func (p *Config) WebAuthnForPasswordless(ctx context.Context) bool {
 }
 
 func (p *Config) WebAuthnConfig(ctx context.Context) *webauthn.Config {
+	id := p.GetProvider(ctx).String(ViperKeyWebAuthnRPID)
+	origin := p.GetProvider(ctx).String(ViperKeyWebAuthnRPOrigin)
+	origins := p.GetProvider(ctx).StringsF(ViperKeyWebAuthnRPOrigins, []string{stringsx.Coalesce(origin, id)})
```
Comment (aeneasr) [inline driver/config/config.go:1370]: I think this is a valid strategy. Alternatively this field can also be polymorph `string | []string` but this is probably a cleaner solution and allows us to remove the origin config at some point.

How would we migrate this in Ory Network?

File: selfservice/flow/recovery/handler.go (hunk 1 above)
Comment (hperl) [reply selfservice/flow/recovery/handler.go:434]: Good catch, thanks!

File: selfservice/strategy/webauthn/js/webauthn.js [hunk 3]
```diff
@@ -1,99 +1,130 @@
...
   if (window.__oryWebAuthnInitialized) {
     return
   }
 
   function __oryWebAuthnBufferDecode(value) {
-    return Uint8Array.from(atob(value), function (c) {
-      return c.charCodeAt(0)
-    });
+    return Uint8Array.from(
+      atob(value.replaceAll("-", "+").replaceAll("_", "/")),
```
Comment (hperl) [inline selfservice/strategy/webauthn/js/webauthn.js:19]: All values are now base64urlencoded by the webauth library.

Comment (hperl) [reply selfservice/strategy/webauthn/js/webauthn.js:19]: The rest of the changes are through prettier.

File: driver/config/config.go (hunk 2 above)
Comment (hperl) [reply driver/config/config.go:1370]: I guess we need a new DB column, read from the old one and then only write to the new one. And we need an UI update, right?

Comment (hperl) [reply driver/config/config.go:1370]: Or we keep the old DB column, and have the origins comma-separated.

**Review by Benehiko — APPROVED** [review]
LGTM

**Comment by hperl** [comment]
Docs PR: https://github.com/ory/docs/pull/1474

File: driver/config/config.go (hunk 2 above)
Comment (aeneasr) [reply driver/config/config.go:1370]: New column 👍

File: driver/config/config.go [hunk 4]
```diff
@@ -1364,14 +1365,17 @@ func (p *Config) WebAuthnForPasswordless(ctx context.Context) bool {
...
```
Comment (aeneasr) [inline driver/config/config.go:1370]: The ID is not a URL iirc but instead the hostname so `ory.sh` for example. The origins however need to have the scheme. So I'm not sure if `id` is a good default.

**Review by aeneasr — APPROVED** [review]

File: driver/config/config.go (hunk 4 above)
Comment (hperl) [reply driver/config/config.go:1370]: I took that behaviour from the config spec: "An explicit RP origin. If left empty, this defaults to `id`."

https://github.com/ory/kratos/pull/3380/files#diff-4a79b8650ca1bb0046229c6603ab7458e7711fc69812f13bbb8fc73b81b833e6R1601

In any case, origin is optional in the config but required for the WebAuthN config.

Comment (aeneasr) [reply driver/config/config.go:1370]: OK SGTM

File: driver/config/config.go [hunk 5]
```diff
@@ -1364,14 +1364,17 @@ func (p *Config) WebAuthnForPasswordless(ctx context.Context) bool {
...
 func (p *Config) WebAuthnConfig(ctx context.Context) *webauthn.Config {
+	id := p.SelfPublicURL(ctx).Scheme + p.GetProvider(ctx).String(ViperKeyWebAuthnRPID)
+	origin := p.GetProvider(ctx).String(ViperKeyWebAuthnRPOrigin)
+	origins := p.GetProvider(ctx).StringsF(ViperKeyWebAuthnRPOrigins, []string{stringsx.Coalesce(origin, id)})
 	return &webauthn.Config{
 		RPDisplayName: p.GetProvider(ctx).String(ViperKeyWebAuthnRPDisplayName),
-		RPID:          p.GetProvider(ctx).String(ViperKeyWebAuthnRPID),
-		RPOrigin:      p.GetProvider(ctx).String(ViperKeyWebAuthnRPOrigin),
-		RPIcon:        p.GetProvider(ctx).String(ViperKeyWebAuthnRPIcon),
+		RPID:          id,
```
Comment (aeneasr) [inline driver/config/config.go:1372]: The `RPID` has to be the hostname - no scheme

The `origin` has to be the scheme+hostname

**Review by aeneasr — CHANGES_REQUESTED** [review]

File: driver/config/config.go (hunk 5 above)
Comment (hperl) [reply driver/config/config.go:1373]: right, fixed.

File: driver/config/config_test.go [hunk 6]
```diff
@@ -1194,6 +1194,51 @@ func TestOAuth2Provider(t *testing.T) {
...
 
+func TestWebauthn(t *testing.T) {
+	ctx := context.Background()
+
+	t.Run("case=multiple origins", func(t *testing.T) {
+		conf, err := config.New(ctx, logrusx.New("", ""), os.Stderr,
+			configx.WithConfigFiles("stub/.kratos.webauthn.origins.yaml"))
+		require.NoError(t, err)
+		webAuthnConfig := conf.WebAuthnConfig(ctx)
+		assert.Equal(t, "https://example.com/webauthn", webAuthnConfig.RPID)
```
Comment (aeneasr) [inline driver/config/config_test.go:1205]: RPID is a domain, not https, no path

https://www.w3.org/TR/webauthn-2/#scope

File: driver/config/stub/.kratos.webauthn.invalid.yaml [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+          - id: github
+            provider: github
+            client_id: a
+            client_secret: b
+            mapper_url: http://test.kratos.ory.sh/default-identity.schema.json
+    webauthn:
+      enabled: true
+      config:
+        rp:
+          id: https://example.com/webauthn
```
Comment (aeneasr) [inline driver/config/stub/.kratos.webauthn.invalid.yaml:81]:
```suggestion
          id: example.com/webauthn
```

File: driver/config/config.go [hunk 8]
```diff
@@ -1364,14 +1364,18 @@ func (p *Config) WebAuthnForPasswordless(ctx context.Context) bool {
 }
 
 func (p *Config) WebAuthnConfig(ctx context.Context) *webauthn.Config {
+	scheme := p.SelfPublicURL(ctx).Scheme
+	id := p.GetProvider(ctx).String(ViperKeyWebAuthnRPID)
+	origin := p.GetProvider(ctx).String(ViperKeyWebAuthnRPOrigin)
+	origins := p.GetProvider(ctx).StringsF(ViperKeyWebAuthnRPOrigins, []string{stringsx.Coalesce(origin, scheme+id)})
```
Comment (aeneasr) [inline driver/config/config.go:1370]:
```suggestion
	origins := p.GetProvider(ctx).StringsF(ViperKeyWebAuthnRPOrigins, []string{stringsx.Coalesce(origin, fmt.Sprintf("%s://%s, scheme, id))})
```

**Review by aeneasr — CHANGES_REQUESTED** [review]

**Review by aeneasr — APPROVED** [review]

**Pull request merged**